llvm-general/src/LLVM/General/Internal/FFI/Attribute.hs

```diff
@@ -126,6 +128,25 @@ foreign import ccall unsafe "LLVM_General_AttrBuilderAddAlignment" attrBuilderAd
 foreign import ccall unsafe "LLVM_General_AttrBuilderAddStackAlignment" attrBuilderAddStackAlignment ::
   Ptr FunctionAttrBuilder -> Word64 -> IO ()
 
+-- The CInt is 0 if the last value is null and 1 otherwise
+foreign import ccall unsafe "LLVM_General_AttrBuilderAddAllocSize" attrBuilderAddAllocSize ::
+  Ptr FunctionAttrBuilder -> Word64 -> CInt -> Word64 -> IO ()
+
 foreign import ccall unsafe "LLVM_General_AttrBuilderAddDereferenceableAttr" attrBuilderAddDereferenceable ::
   Ptr ParameterAttrBuilder -> Word64 -> IO ()
-                                              
+
+foreign import ccall unsafe "LLVM_General_AttrBuilderAddDereferenceableOrNullAttr" attrBuilderAddDereferenceableOrNull ::
+  Ptr ParameterAttrBuilder -> Word64 -> IO ()
+
+foreign import ccall unsafe "LLVM_General_AttributeGetAllocSizeArgs" attributeGetAllocSizeArgs' ::
+  FunctionAttribute -> Ptr Word64 -> Ptr Word64 -> IO CInt
+
+attributeGetAllocSizeArgs :: FunctionAttribute -> IO (Word64, Maybe Word64)
+attributeGetAllocSizeArgs attrBuilder =
+  alloca $ \x ->
+  alloca $ \y ->
```

Comment:
This kind of use of alloca/peek is what the DecodeAST stuff is for. This level of the process belongs up out of the FFI directory, which is as much just-for-imports as I can get it. One exception I've made is currying, where the general Encode/Decode mechanics sometimes work better with partially uncurried types which the imports can't express (e.g. pairing successive arguments into (pointer,length) pairs to pass arrays.